chore: ENSAdmin - fix data model for `Latest Indexed Registrations` panel

Y3drk opened the pull request:
Solves Issue #548 .

Y3drk commented:
Overall, I have a feeling that this refinement is quite redundant, and could be replaced with just some additional documentation. 

Maybe I didn't fully lean into the requirements described in the Issue, but I don't know what else to add there.

Awaiting feedback :)

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,10 +15,35 @@ export interface Domain {
+ * owner - Address -  owner's address
+ * wrappedOwner - Address - optional - wrapped owner's address
+ **/
+export interface LatestRegistration {
```
```suggestion
export interface Registration {
```

Y3drk replied:
@lightwalker-eth 

This change conflicts with another type used in that panel, that's named exactly `Registration` in **\recent\registrations\types.ts**...Maybe we could unify them?

lightwalker-eth replied:
These should be unified into a single type.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,10 +15,35 @@ export interface Domain {
 
+/**
+  A "pure" version of LatestRegistrationResult that streamlines data usage in RecentRegistrations component.
+ @params:
```
We shouldn't document params here like this.

Instead, the best practice is to document each field directly where each field is defined.

```
/**
 * The registered ENS name
 */
name: string;
```

Make sure you always document these with /** style comments and not // style comments. Otherwise IDEs don't read it properly.

(edit: GitHub is messing up the alignment of th

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -46,6 +51,17 @@ async function fetchRecentRegistrations(baseUrl: URL): Promise<RecentRegistratio
+      domain: LatestRegistrationResult;
+    }) => ({
+      ...latestRegistration,
+      domain: toLatestRegistration(latestRegistration.domain),
```
A domain is not a registration. We're getting the ideas all mixed up here.

lightwalker-eth reviewed (changes requested):
@Y3drk Thanks for updates. Reviewed and shared feedback.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
+  createdAt: number;
+
+  /**
+   * a UNIX timestamp in seconds of when will the current owner's tenure end
```
```suggestion
   * a UNIX timestamp in seconds when the registration is scheduled to expire
```

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
+  /**
+   * owner's address
+   */
+  owner: Address;
```
Suggest to rename this to `ownerInRegistry`.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
+  owner: Address;
+
+  /**
+   * wrapped owner's address
```
```suggestion
   * The owner according to the ENS NameWrapper. If undefined, the domain associated with the registration is unwrapped (not in the ENS NameWrapper).
```

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
+  expiryDate: number;
+
+  /**
+   * owner's address
```
```suggestion
   * The "true" owner of the domain in the ENS Registry.
```

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
+ **/
+export interface LatestRegistration {
+  /**
+   * recently registered ENS name
```
```suggestion
   * The registered ENS name
```

This data model should not be specific to a "recent registration". It should be general to any "registration"

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -1,9 +1,9 @@
+ * The data model returned by a GraphQL query for the latest registrations.
  */
-export interface Domain {
+export interface LatestRegistrationResult {
```
This definition should not be here. It should be in `apps/ensadmin/src/components/recent-registrations/hooks.ts` and it should NOT be exported from that file. It should be an internal implementation detail of the hook and not leaked outside of that.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
+  /**
+   * Data associated with a Registration event.
+   */
+  registration: LatestRegistration;
```
This field should not exist. There shouldn't be both a `LatestRegistration` and a `Registration` type. These ideas should be merged together into a single type: `Registration`.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
+  /**
+   * a UNIX timestamp in seconds of when will the current owner's tenure end
+   */
+  expiryDate: number;
```
```suggestion
  expiresAt: number;
```

It's confusing to call this a "date" as that could be interpreted as just a date ex: "Jan 1, 2025" rather than a full timestamp.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
+  /**
+   * a UNIX timestamp in seconds of when the name was registered by its current owner
+   */
   registrationDate: string;
```
We shouldn't be mixing types used for timestamps. We should use the same type everywhere for timestamps. Please suggest which we should use.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
+  /**
+   * a UNIX timestamp in seconds of when will the current owner's tenure end
+   */
   expiryDate: string;
```
We shouldn't be mixing types used for timestamps. We should use the same type everywhere for timestamps. Please suggest which we should use.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
 }
 
+/**
+ * Data about 5 most recent registrations
```
A few issues here:

1. This shouldn't make any reference to the number 5. The number of records contained should be any number.
2. This type should not live in this file. It should live inside the hooks.ts file and it should not be exported out of that file. It is an internal implementation detail.
3. The `registrations` field here should NOT be a `Registration`. It should be a `RegistrationResponse`. You're mixing up ideas. Please 100% separate the internal implementation details of GraphQL response schemas (which are ugly and shit due to a number of factors) from the 100% clean and perfect data model we want to use across ENSAdmin for Registrations. The ugly implementation details of GraphQL response schemas must be TRAPPED inside a single layer and not allowed to leak out anywhere else across our code. Only the 100% perfect data model of a `Registration` should spread across our code.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -15,12 +15,59 @@ export interface Domain {
+  /**
+   * wrapped owner's address
+   */
+  wrappedOwner?: Address;
```
Suggest to rename this to `ownerInNameWrapper`.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -65,3 +81,20 @@ export function useRecentRegistrations(searchParams: URLSearchParams) {
+ * @param graphQLResponseData - LatestRegistrationResult - GraphQL compatible representation of recently registered domain, dubed 'domain' in the query
+ * @returns fetched data in new, streamlined LatestRegistration format
+ */
+export function toLatestRegistration(
```
This function should not be exported. It should be an internal implementation detail.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -1,6 +1,11 @@
+} from "./types";
 
 /**
  * Fetches info about the 5 most recently registered .eth domains that have been indexed.
```
This function should not return a `RecentRegistrationsResponse` this is leaking ugly GraphQL shit all over our code. It needs to return `Promise<Registration[]>`

lightwalker-eth replied:
It's not a good design to make specific reference to a number such a "5" here. The number of registrations to query should be a param passed into the function.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -46,6 +51,17 @@ async function fetchRecentRegistrations(baseUrl: URL): Promise<RecentRegistratio
   }
 
   const data = await response.json();
+  data.data.registrations = data.data.registrations.map(
```
We should be completely transforming a `RegistrationResult` to a `Registration`.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/components.tsx:
```diff
@@ -135,18 +136,19 @@ const NAME_WRAPPER_ADDRESS = "0xd4416b13d2b3a9abae7acd5d6c2bbdbe25686401";
+ * @param registration - Extended data associated with Registration event
  */
-function getTrueOwner(owner: { id: Address }, wrappedOwner?: { id: Address }) {
+function getTrueOwner(registration: Registration) {
```
@Y3drk This function shouldn't live in this file.

1. The logic here should move into the hook.
2. This function should not be exported. It is an internal implementation detail.
3. The `Registration` type should have another field added: `owner` (please see my other comments suggesting renaming of the existing `owner` field to `ownerInRegistry`. This NEW `owner` field serves the purpose of identifying the effective owner in a simple way.
4. Whenever a `Registration` is built by the DAL, the DAL should have the responsibility of properly setting this NEW `owner` field by calling this internal implementation detail helper function.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/components.tsx:
```diff
@@ -235,10 +237,7 @@ export function RecentRegistrations() {
-                            registration.domain.owner,
-                            registration.domain.wrappedOwner,
-                          )}
+                          address={getTrueOwner(registration)}
```
```suggestion
                          address={registration.owner}
```

Please see my related comment above for more details. When we get all this code right, this suggestion shows how the final result should look here.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/components.tsx:
```diff
@@ -135,18 +136,19 @@ const NAME_WRAPPER_ADDRESS = "0xd4416b13d2b3a9abae7acd5d6c2bbdbe25686401";
-    return getAddress(wrappedOwner.id);
+  if (
+    registration.registration.wrappedOwner &&
+    isAddressEqual(registration.registration.owner, NAME_WRAPPER_ADDRESS)
```
FYI: Here what we're really doing is we're checking if the `ownerInRegistry` is equal to the `NAME_WRAPPER_ADDRESS`. If so, then `ownerInNameWrapper` MUST be defined (otherwise we should throw an exception as our expectations are violated) and the `ownerInNameWrapper` should be considered the effective owner.

Y3drk commented:
# Please read before review

I've implemented the vast majority of @lightwalker-eth 's suggestions, but some things weren't 100% clear to me, so I made some assumptions, which I will describe in detail later. Also, I have several questions regarding those points of uncertainty that I will also write below.

## Assumptions
1. Latest feedback included a suggestion to unify the type that we associate with various `timestamps` in the `Registration` type. I've chosen to go with `string` type, since this is the form these fields have when returned by a GraphQL query, and all of the helper functions that were previously written expect the input timestamp to be in string format.
2. One of the suggestions was to unify `LatestRegistration` and `Registration` types and to do that I made several assumptions:
    a. I injected the `LatestRegistration` type into `Registration` and **flattened** it, assuming that we would want to avoid nested objects.
    b. The aforementioned unification resulted in two separate fields having the same name: `expiresAt`. After some investigation I ruled out reduction to a single field since the values were not identical and decided to rename one of the fields. To my knowledge, when returned as a GraphQL query result, the `expiryDate` field that is returned inside the "domain" object is the value of the "outer" `expiryDate` object plus the [**grace period**](https://docs.ens.domains/terminology#grace-period) set by ENS to 90 days, hence I renamed this field to the `expiresAtWithGracePeriod`.
    c. To avoid confusion about the meaning of `createdAt` field, I renamed it to `domainCreatedAt` so that it's clear to what item it refers to
    d. As suggested I renamed previous owner fields and added a new one: `owner`, but I wasn't certain about its description.
    
Here is how the unified `Registration` type looks:
```
export interface Registration {

  registeredAt: string;

  expiresAt: string;

  name: string;

  domainCreatedAt: string;

  expiresAtWithGracePeriod: string;

  ownerInRegistry: Address;

  ownerInNameWrapper?: Address;

  owner: Address;
}
```
Fields' descriptions were excluded and can be seen in the PR's files.

3. I didn't implement the suggestion about making a hardcoded limit of recent registrations (currently set to 5) a function input, but I plan on doing it.

## Questions
1. Is the proposed type fine with you?
2. Questions about new `Registration` type:
    a. Is the proposed, unified type acceptable, should we go back to its nested version? 
    > Personally I prefer it the way it is now
    
    b. The response from GQL has one more field `labelName` that we don't use. Should we add it to the `Registration` type?
    c. Do we want to remove any of the current fields from this type?
    > As of yet we never use `domainCreatedAt` and `expiresAtWithGracePeriod` fields, but I suppose that, they might come in handy in the future
    

3. Do we want the users to be able to set the limit of registrations themselves or we just want it as a function parameter for the sake of clean-code?

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/types.ts:
```diff
@@ -1,26 +1,48 @@
+// TODO: fix the order of the fields for improved clarity
+export interface Registration {
+  /**
+   * a UNIX timestamp in seconds of when the domain was registered by its current owner
```
When we write comments about a field, we should read it back to ourselves as if we were a CPU.

This is not actually a UNIX timestamp, it is a string formatted as an integer that represents a UNIX timestamp. Please take special note of how we should document our ideas for accuracy.

Suggestion is for this to just be a `Date` however, so the feedback above is just general advice for the future.

lightwalker-eth reviewed (changes requested):
@Y3drk Thanks for updates. Reviewed and shared feedback :+1:

Y3drk commented:
# Before review (06/17/25)

## New changes
* I've applied all @lightwalker-eth's suggestions from all previous PRs, including:
    * Refactoring and moving helper functions to `recent-recgistrations/utils.tsx` file --> these functions could be moved to `/components/ui` and further refactored if deemed necessary. Need to decide on their final location
    * Creating a new parsing function `transformTimestamp` in `recent-recgistrations/hooks.ts` that allowed to move the validation of the data completely into the `hooks.ts` file, thus making the validation of the GraphQL response an internal app process
    * Investigation about the inclusion of a grace period in one of the `expiresAt` fields (see comment in the review above)
    * Changing & unifying all timestamp-related fields in the `Registration` type to `Date` objects.
    * Renaming some internal types and improvements of docstrings in general
* Fixed edge case for very short durations of registrations, where the component previously displayed `0 months`. Now for these cases, the returned value is `less than a month`. If preferred, we can make it more detailed.

## Note about uncertainties & questions
In cases when I wasn't sure what to do, I left a comment right under the review suggestion. Would appreciate having a look at them first if the changes aren't satisfactory. 

For now, I don't think I have any additional questions, besides those mentioned above.

Awaiting feedback and thanks in advance :)

lightwalker-eth commented on apps/ensadmin/src/components/identity/index.tsx:
```diff
@@ -52,7 +52,7 @@ export function ENSName({
 
```
GitHub won't let me put this comment on the correct line. This comment is actually for several lines above:

```
const ensAppUrl = `https://app.ens.domains/${address}`;
```

Please see my other comment about how this logic is not correct as it needs to consider the ENS deployment chain of the connected ENSNode.

lightwalker-eth commented on apps/ensadmin/src/components/identity/index.tsx:
```diff
@@ -52,7 +52,7 @@ export function ENSName({
+    return <IdentityPlaceholder showAvatar={showAvatar} className={className} />;
   }
 
   return (
```
GitHub won't let me put this comment on the correct line. This comment is actually for several lines below:

```
            <AvatarImage
              src={`https://metadata.ens.domains/mainnet/avatar/${ensName}`}
              alt={ensName}
            />
```

This logic is not correct:
1. It should not always use `mainnet` in the image path. Depending on the ENS deployment chain it should use holesky or sepolia.
2. The logic needs to consider the case that the ENS deployment chain is the `ens-test-env`. In this case there's no ENS Metadata Service we can talk to and therefore we should always just display the AvatarFallback.

Y3drk commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -1,17 +1,95 @@
+ *
+ * @param registrationResult - Data associated with Registration event
+ */
+function getTrueOwner(registrationResult: RegistrationResult) {
```
Applied, but left the comment open for future reference

Y3drk commented on apps/ensadmin/src/components/recent-registrations/utils.tsx:
```diff
@@ -0,0 +1,75 @@
+/**
+ * Helper function to calculate duration length in years or months
+ */
+const formatDuration = (beginsAt: Date, endsAt: Date) => {
```
@lightwalker-eth 

Noted. From now on, the `formatDuration` function will be fully replaced with the `formatDistanceStrict` from `date-fns`. 

About leveraging functions options &rarr; As far as I am aware, the default version of this function is perfect for us since it doesn't add any suffixes or force any specific unit, hence it should stick to displaying the duration in the largest unit possible (for our use-case usually years, sometimes months) just like we did until now.

Y3drk replied:
This suggestion is resolved but kept open for future reference

Y3drk commented on apps/ensadmin/src/components/recent-registrations/utils.tsx:
```diff
@@ -0,0 +1,75 @@
+  const years = differenceInYears(endsAt, beginsAt);
+
+  // If less than a year, show months instead
+  if (years === 0) {
```
I'll keep that in mind, but this particular case was solved by replacing our custom function with the `formatDistanceStrict` from `date-fns` library

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -1,17 +1,86 @@
+}
+
+/**
+ * Numeric string representing a Unix timestamp in seconds.
```
```suggestion
 * An integer value (representing a Unix timestamp in seconds) formatted as a string.
```

This states the idea more clearly. It's not so clear what a "Numeric string" is.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -1,17 +1,86 @@
+
+function transformTimestamp(timestamp: UnixTimestampInSeconds): Date {
+  try {
+    return new Date(parseInt(timestamp) * millisecondsInSecond);
```
Please write unit tests for `transformTimestamp`. I believe you'll see that the logic on this line won't throw an exception on invalid `UnixTimestampInSeconds` values as you expect.

Please see how we've defined other unit tests for examples.

As I understand, this is the correct pattern in Javascript:

```
const d = new Date("foo"); // NOTE: this won't throw an exception!
if (Number.isNaN(d.getTime())) {
  // invalid date -- here's where you actually throw an exception.
}
```

Y3drk replied:
# Update (from the sync call) 06/23/25

The function will be moved to the `utils.tsx` file and exported. The related type alias will probably have to be exported as well

Then its test will simply be moved to a separate `utils.test.ts` file as we would usually do with vitest tests

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -1,17 +1,86 @@
+  return getAddress(registrationResult.domain.owner.id);
+}
+
+function transformTimestamp(timestamp: UnixTimestampInSeconds): Date {
```
```suggestion
function unixTimestampToDate(timestamp: UnixTimestampInSeconds): Date {
```

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -1,17 +1,86 @@
+    expiresAt: transformTimestamp(registrationResult.expiryDate),
+    name: registrationResult.domain.name,
+    ownerInRegistry: registrationResult.domain.owner.id,
+    owner: getEffectiveOwner(registrationResult),
```
When building an instance of object X, it's nice to list the fields in the same order they are defined for object X.

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -1,17 +1,86 @@
+}
+
+/**
+ * Fetches info about most recent registrations that have been indexed.
```
```suggestion
 * Fetches info about the most recent registrations that have been indexed.
```

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -46,22 +115,24 @@ async function fetchRecentRegistrations(baseUrl: URL): Promise<RecentRegistratio
+    queryFn: () => fetchRecentRegistrations(ensNodeURL, maxResults),
     throwOnError(error) {
-      throw new Error(`Could not fetch ENSNode data from '${ensNodeUrl}'. Cause: ${error.message}`);
+      throw new Error(`Could not fetch ENSNode data from '${ensNodeURL}'. Cause: ${error.message}`);
```
```suggestion
      throw new Error(`Could not fetch recent registrations from '${ensNodeURL}'. Cause: ${error.message}`);
```

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -46,22 +115,24 @@ async function fetchRecentRegistrations(baseUrl: URL): Promise<RecentRegistratio
+ * Hook to fetch info about most recently registered domains that have been indexed.
+ *
+ * @param ensNodeURL The URL of the selected ENS node instance.
+ * @param maxResults number of the maximal number of latest registrations to be retrieved by the query
```
```suggestion
 * @param maxResults the max number of recent registrations to retrieve
```

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -1,17 +1,86 @@
+ */
+function getEffectiveOwner(registrationResult: RegistrationResult): Address {
+  // Only use wrapped owner if the owner is the NameWrapper contract
+  if (isAddressEqual(registrationResult.domain.owner.id, NAME_WRAPPER_ADDRESS)) {
```
There's unnecessary nesting of `if` statements here. Please see this article that talks about the best practice: https://www.linkedin.com/pulse/avoiding-nested-statements-guard-clauses-danny-olsen-a1e8f/

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/hooks.ts:
```diff
@@ -1,17 +1,86 @@
+    registeredAt: transformTimestamp(registrationResult.registrationDate),
+    expiresAt: transformTimestamp(registrationResult.expiryDate),
+    name: registrationResult.domain.name,
+    ownerInRegistry: registrationResult.domain.owner.id,
```
Please note: You haven't validated that the following fields are valid address values using the `getAddress` function:
- `registrationResult.domain.owner.id`.
- `registrationResult.domain.wrappedOwner?.id`. -- take care to ensure getAddress is only called here if this address value is defined.

lightwalker-eth reviewed (changes requested):
@Y3drk Thanks for updates :+1: Identified some other issues for us to fix here. Thanks

lightwalker-eth commented on apps/ensadmin/src/components/recent-registrations/utils.tsx:
```diff
@@ -51,3 +52,21 @@ export function Duration({
+export type UnixTimestampInSeconds = string;
+
+/**
+ * Transforms a timestamp formatted as a string to a Date object.
```
```suggestion
 * Transforms a UnixTimestampInSeconds to a Date.
```

Best to build upon definitions you've already invested in.

lightwalker-eth reviewed (changes requested):
@Y3drk Overall code review looks good now, just 1 very trivial feedback, but we must have missed something. Please check the Vercel preview environment:

![CleanShot 2025-06-23 at 18 49 15](https://github.com/user-attachments/assets/47af7dc4-01a3-4155-a1f1-0cf443d9f6e8)
